src/ui/Controllers/CourseController.cs

```diff
@@ -14,25 +14,36 @@ namespace GovUk.Education.ManageCourses.Ui.Controllers
     [Route("organisation")]
     public class CourseController : CommonAttributesControllerBase
     {
-        private readonly ManageApi _manageApi;
+        private readonly IManageApi _manageApi;
 
-        public CourseController(ManageApi manageApi)
+        public CourseController(IManageApi manageApi)
         {
             _manageApi = manageApi;
         }
         [Route("{instCode}/course/{accreditingProviderId=self}/{ucasCode}")]
         public async Task<IActionResult> Variants(string instCode, string accreditingProviderId, string ucasCode)
         {
+            Validate(instCode, accreditingProviderId, ucasCode);
+
             var course = await _manageApi.GetCoursesByOrganisation(instCode);
 
             var providerCourse = "self".Equals(accreditingProviderId, StringComparison.InvariantCultureIgnoreCase)
                 ? course.ProviderCourses.First(c => String.IsNullOrEmpty(c.AccreditingProviderId))
                 : course.ProviderCourses
                     .First(c => c.AccreditingProviderId.Equals(accreditingProviderId, StringComparison.InvariantCultureIgnoreCase));
-            var courseDetail = providerCourse.CourseDetails.First(x => x.Variants.Select(v => v.UcasCode == ucasCode).Any());
+
+            var courseDetail = providerCourse.CourseDetails.FirstOrDefault(c => c.Variants.Any(v => v.UcasCode == ucasCode));
+
+            if (courseDetail == null)
+            {
+                throw new Exception("Unexpected error: course should not be null");
```

Comment:
suggestion:

    throw new InvalidOperationException($"Course variant with ucas code '{ucasCode}' not found");

https://stackoverflow.com/questions/1453200/exception-for-missing-data

Minor.

Reply:
Good idea. Done